### PR TITLE
[WIP] Hot reloading and redesign

### LINKED ISSUE #47: Autostart the node
Users might wonder why they have to start the node. “Did I not just start IPFS already?”

### PR DESCRIPTION
**Pull request by Dignifiedquire**
> **Last updated** 11.10.2015 00:25

## Todo list

- [x] Add travis
- [x] Implement react module reloading
- [x] Add better debugging capabilities to electron
- New styles for
  - [x] Start Screen
  - [x] Fany loading animation
  - [x] Profile Screen
    - [x] logo in menu title `[  ] IPFS` (Ref https://github.com/ipfs/logo/issues/6)
    - [x] hide settings for now
    - [x] Auto start node (Closes #47 )
  - [x] Welcome Process
    - [x] advanced options screen
    - [x] move path into advanced
    - [x] add key choice into advanced. (make algo a dropdown, with only RSA but say "ECC and more coming soon". it's an important detail)
    - [x] Close application on setup cancel
    - [x] Auto start application and show window on finished init
  - [x] File Upload
- [ ] Test packaging with new build setup (blocked by https://github.com/jbenet/node-subcomandante/issues/1)
- [ ] Documentation
  - [ ] Components
  - [ ] File structure
  - [ ] Used libraries
- [x] Remove code cruft
- [x] Add proper file and console logging
- [x] React 0.14

## Progress images

Making progress:
ipfs-hot-reload

### CODE REVIEW

**Comment by diasdavid** [comment]
This is really sweet @Dignifiedquire! it will help not only the electron app, but also bringing this front end dev best practices to other projects like the WebUI and StarLog

**Comment by masylum** [comment]
:muscle: babel!

**Comment by Dignifiedquire** [comment]
Made some progress today and there is now a pretty start screen, with custom fonts and icons:

screen shot 2015-09-21 at 01 43 38

**Comment by jbenet** [comment]
@Dignifiedquire looking good! solid progress. 

Though, let's please sketch all the screens out first. we want to provide a cohesive look, and one that addresses some plans for this app that are probably not really documented anywhere yet (we've discussed them on irc and other channels, but not really listed). I can spend some time working with you on this. btw, take a look at these, too: https://github.com/ipfs/electron-app/issues/3#issuecomment-111764862 -- not set on this one either, but it's a different roll that incorporates useful information shown even when offline.

**Comment by jbenet** [comment]
@Dignifiedquire when do you have 30min next to discuss some plans? (or maybe i can post thoughts in form of sketches here in this repo)

**Comment by jbenet** [comment]
(Anyway, super excited about this progress :+1: :+1: :+1: :+1: )

**Comment by Dignifiedquire** [comment]
@jbenet I'm around in irc, just ping me

**Comment by Dignifiedquire** [comment]
@jbenet here is the mockup I made

electron-app-01-v2

**Comment by Dignifiedquire** [comment]
Enough for tonight :) 

screen shot 2015-09-21 at 21 21 23

**Comment by Dignifiedquire** [comment]
ipfs-progress

**Comment by jbenet** [comment]
Looking good!

File: app/controls/open-settings.js [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 
 function openSettings () {
-  var settingsWindow = new BrowserWindow(config.window)
+  const settingsWindow = new BrowserWindow(config.window)
+
+  if (process.env.NODE_ENV === 'production') {
+    settingsWindow.loadUrl('file://' + path.resolve(__dirname, '../settings.html'))
+  } else {
+    settingsWindow.loadUrl('http://localhost:3000/settings.html')
+  }
```
Comment (masylum) [inline app/controls/open-settings.js]: this seems a recurring pattern in this PR. Worth abstracting it?

Comment (diasdavid) [inline app/controls/open-settings.js]: Thanks for catching that @masylum , all that is config should go to the config.js file https://github.com/ipfs/electron-app/blob/master/app/config.js

Comment (Dignifiedquire) [inline app/controls/open-settings.js]: done

**Comment by Dignifiedquire** [comment]
Couldn't work on this in the last week since I was sick lying in bed :mask: But getting back to it slowly

**Comment by Dignifiedquire** [comment]
@jbenet @diasdavid I'm thinking of dropping the switch for setting the key size and just default to `4048`. I just feel that most new users will have no idea what the difference is anyway and it's just another hurdle in getting setup and things running. Do you have any objections/thoughts?

**Comment by diasdavid** [comment]
@Dignifiedquire sorry to hear you were sick, hope you are recovering well! 

When there is no IPFS node already set up, electron-app opens a welcome screen to set those defaults, we can leave a sane default pre-selected of 4096, but why removing the option to choose?

**Comment by jbenet** [comment]
We'll also add algorithms later (like ECC with specific curves (ed25519)). Maybe there could be an "advanced" toggle that hides the options? Not sure



—

**Comment by Dignifiedquire** [comment]
@diasdavid Thanks, getting better every day now.

The main reason here is simplicity. I expect a lot of users that start using ipfs through this app will simply not know what these numbers actually mean, so they have no way of making a good choice, but rather the get scared and might think "Oh god this too complicated/techy/.. for me" and we loose them at this very first point. And even if they continue they will have the nagging in their mind, "maybe I choose the wrong number?"

If you compare it with the following screen, there is only one simple step so people don't have to think about it and do not get scared away (hopefully). 

screen shot 2015-09-26 at 22 50 32

screen shot 2015-09-26 at 22 55 03

If we actually have more options, like different algorithms we could start thinking about adding an advanced section, hidden by default where you can specify different things about the key generation.

**Comment by diasdavid** [comment]
What about having a 'advanced options' button for power users and a continue with sane defaults for non power users ?

**Comment by jbenet** [comment]
+1 for advanced options. :-1: for not having options.

could also be a `> Advanced Options` toggle for a collapsed div or something

**Comment by jbenet** [comment]
also, while i do really like how "all the awesomeness that is IPFS" sounds :), maybe we should say "the IPFS content" or "IPFS data", as we will have users not very receptive to the former.

**Comment by jbenet** [comment]
:+1: for having the window be larger than the file picker popup, that's a really good design choice.

**Comment by jbenet** [comment]
and :+1: in general for the style. (sorry for million-posts)

**Comment by jbenet** [comment]
Remainder TODOs here:
- [ ] init screen
   - [ ] advanced options tab
   - [ ] move path into advanced
   - [ ] add key choice into advanced. (make algo a dropdown, with only RSA but say "ECC and more coming soon". it's an important detail)
- [ ] menu screen
  - [ ] logo in menu title `[  ] IPFS`
  - [ ] hide settings for now
  - [ ] figure out if "console" and "browser" can be more clear (a bit confusing atm)
- [ ] file upload
  - [ ] re-enable file upload
  - [ ] add progress bars for file upload

**Comment by Dignifiedquire** [comment]
New mockups for the welcome screen according to the discussed ideas. 

welcome-initial-v2

welcome-customization

**Comment by jbenet** [comment]
@Dignifiedquire these look awesome!!! very very cool!

- have you tried the buttons as just a thick outline? like this: https://dribbble.com/shots/1762899-The-Mars-Route-by-SpaceX/attachments/287139
- is the image CC or stock photo we can buy?
- maybe "Advanced" instead of "Customize"? it's a more common thing to see
- are the form fields in advanced + buttons just css on top of straight bootstrap? we need to preserve the ease of modification of all these screens. that's worth more than really nice-looking things. i.e. it should still take someone a couple min to be able to add a field.

**Comment by jbenet** [comment]
also, the designer who made https://dribbble.com/shots/1762899-The-Mars-Route-by-SpaceX/attachments/287139 is a friend, and he could likely do some work for us.

**Comment by Dignifiedquire** [comment]
> is the image CC or stock photo we can buy?

The image is cc from https://unsplash.com/

> maybe "Advanced" instead of "Customize"? it's a more common thing to see

Sure I don't mind either way

> are the form fields in advanced + buttons just css on top of straight bootstrap?

We don't use bootstrap anymore, but yes coding this is very straight forward

**Comment by Dignifiedquire** [comment]
> have you tried the buttons as just a thick outline?

Yes I tried it, but these kind of buttons do not work as well in an application as they do on a website.

**Comment by jbenet** [comment]
> We don't use bootstrap anymore, 

wait, why?

> but yes coding this is very straight forward

can i get an example?

**Comment by Dignifiedquire** [comment]
> wait, why?

Because it's a lot of code which we don't need and it makes just overly complicated if you want to customize things in the way we are doing.

> can i get an example?

Sure: https://github.com/Dignifiedquire/electron-app/blob/settings-hot/app/js/components/setup.js#L114-L122

**Comment by jbenet** [comment]
i understand that bootstrap is not liked by more pro web devs, but it makes a _huge_ difference for being able to prototype things, and modify them as a non-web dev person. it is a very easy to use UX language, and should not be divested of as just a cumbersome framework. not using it anymore is something i'm not comfortable with.

**Comment by Dignifiedquire** [comment]
I've stated at the very top of the issue that I dropped bootstrap :( It was one of the first things I did.

**Comment by jbenet** [comment]
> Because it's a lot of code which we don't need and it makes just overly complicated if you want to customize things in the way we are doing.

the problem that emerges it that now, only you know how to modify this stuff. :( 

> I've stated at the very top of the issue that I dropped bootstrap :( It was one of the first things I did.

really? i don't see it :/ sorry if i missed it, i would've been able to address it

**Comment by Dignifiedquire** [comment]
New todo: Properly plan and **document** component and styling structure so @jbenet can sleep better and it doesn't matter if I get hit buy a bus.

**Comment by jbenet** [comment]
IRC discussion for posterity 

```irc
14:44 <dignifiedquire> jbenet: lets talk here, that’s easier
14:44 <dignifiedquire> rather than filling up the pr 
14:44 <•jbenet> dignifiedquire: sire
14:44 <•jbenet> sure*
14:45 <•jbenet> dignifiedquire: i didn't see the earlier mention about bootstrap. i understand the desire to break out and do something really cool and custom. totally get that. but the problem is that now only you know how to modify stuff, and it will take a lot of work for someone to get up to speed on custom styling.
14:45 <•jbenet> the value of bootstrap (and derived things) is defining the closest thing we have to a modular UI language in the web.
14:46 <•jbenet> it's like a UI/UIKit for HTML5.
14:46 <•jbenet> bootstrap is actually very, very hackable. it's not difficult to achieve the styling you want on top of it.
14:47 <dignifiedquire> Yes I understand that, but when you are using react you are not constrained to using only bootstrap anymore because you just write your own modular components that you can reuse everyhwere without having to know bootstrap or a lot about the system
14:47 <•jbenet> it can be done without blowing up all the valuable utility we get from having a common language.
14:47 <dignifiedquire> for example, if I have a finished style for an input, you just have to write sth like <Input type=“text” value={this.state.myVal} /> and be done with it
14:47 → shoman5 joined (~shoman5@46.166.190.182)
14:48 <•jbenet> maybe i'm missing something -- afaik, the modular components in react, all the css in them, is still embedded in the same page, still follows styling rules, and should fit a common application theme. (i.e. the best modules will delegate styling to the top of the application, which can redefine and chnage the style by changing some top level css rules)
14:48 <dignifiedquire> no the styling is not defered to the top level anymore, they are defined where they are needed
14:49 <•jbenet> well, suppose that styling needs to be redone in the future.
14:49 → NeoTeo joined (~NeoTeo@6167030-cl69.boa.fiberby.dk)
14:49 <dignifiedquire> except for global resets, and defaults, like font color et
14:49 <•jbenet> now we'd have to step into every single module and modify it.
14:49 <•jbenet> instead of modifying global styling rules.
14:50 → Eudaimonstro joined (~steve@108-255-43-22.lightspeed.irvnca.sbcglobal.net)
14:53 <dignifiedquire> yes, but that the thing is that for real style changes what you would have is a file/section in a gigantic file where you say: here are all styles for inputs, so when you want to change the whole style you have to go into each of these sections and customize them. If you instead take the styles and put them right in the input component you can just go there and customize it there. 
14:53 <dignifiedquire> Which ends being much easier and straight forward because you have a small scope to care about (just the input) and you see immediately the html parts of it. At the end it’s the same argument why react puts the templates inside the script file because these things actually belong together and you have to touch the styling code anywhere.
14:53 <dignifiedquire> There is of course a place for configurable things and globals/configuration files for things like primary color, font family, etc
14:55 <•jbenet> sure that makes sense, i dont see why this is incompatible with bootstrap-- for example, you can imagine a reactified bootstrap that puts all the components into its own react module, following the pattern you describe.
14:55 <dignifiedquire> also, https://speakerdeck.com/vjeux/react-css-in-js for a little bit better wording of what I’m saying
14:56 <dignifiedquire> Yes and there is a port of bootstrap components for react, which you can use of course
14:56 <•jbenet> but that still is customizable to a great degree by a few global rules (yep, like color schemes all over) that often apply to a whole family of components, not just one
14:56 <dignifiedquire> the thing is why I didn’t do that, is that we don’t need 98% of that code 
14:56 <•jbenet> the concern i have is more around the ability to use these components. like for example, if i want to go and add something new, like a textarea, i no longer can easily.
14:56 <dignifiedquire> why not?
14:57 <•jbenet> what component would i use? are you going to make a full library of custom components as extensive as bootstraps?
14:57 <•jbenet> i cant use bootstrap's as now the style would be all off and wonky
14:58 → Yeelk joined (4728b0aa@gateway/web/freenode/ip.71.40.176.170)
14:59 <dignifiedquire> That depends on your needs, but if you just need a textarea you just put in the default html textarea, hook it up to react as needed and add a style depending on where you put it. 
15:00 <dignifiedquire> If the element wasn’t used before yes a new style definition has to be done, but that is a good thing because one should think about that when adding a new UI element and how it fits into the current interface. 
15:01 <•jbenet> i think what would happen is that people not deep into frontend dev would no longer be able to modify the UX of the codebase. it's already hard as is, but gets increasingly harder. sorry, i dont mean to complain
15:02 <dignifiedquire> well they would need to know some basic html, css and react yes, but that was the case before as well wasn’t it?
15:02 <•jbenet> the misalignment i'm trying expess is that, while i understand that it is more cumbersome to follow the constraints imposed by bootstrap, the important thing is to keep very long-term maintainable and malleable components, without much overhead to using them. like, there's not guarantees at all that this new style will be the right thing in the long term, or
15:02 <•jbenet> that you'll even want to keep contributing to it in the future. Lots of PRs in go-ipfs introduce abstractions what are highly custom and not very maintainable, and we have to fight hard to keep things as easy to maintain for as broad of a base as we can.
15:04 <•jbenet> like, even one of Go's fundamental values is reducing that sort of complexity, coming up with "idiomatic Go" as a thing is to ensure as many people as possible stick to "one way of doing things" because as the months and years go by, and as dozens of people come and go, the codebases stay and evolve little bits by bits.
15:05 <dignifiedquire> Yes, which I totally understand. Of course you want this to be workable longterm, which is for me the reason I’m against bootstrap. I’m working on a project right now where bootstrap was used in the beginning for the exact same reasons (no react though, just jquery + bootstrap) and this projects is now > 2 years old and it’s horrible, you have to work around bootstrap all the time when trying to fix things and change 
15:05 <dignifiedquire> the design when it doesn’t fit into the bootstrap scheme of things
15:07 <•jbenet> is that misuse of bootstrap? in my experience it's always been relatively easy to modify and re-style. way, way easier than the customized layouts pre-bootstrap (i've been dabbling with html5 for a long time and remember the horrible pain of getting UIs right before bootstrap)
15:08 ⇐ Bedlamb quit (~Floophead@71.58.95.78) Quit: Leaving.
15:08 <•jbenet> (and again, here im using "bootstrap" as one instance of a "uikit-style-thing", a common language that's well understood, well supported, comprehensive and robust, and well tested across a variety of devices. bootstrap is so far the most popular of these, but ofc others exist.
15:08 → Bedlamb joined (~Floophead@71.58.95.78)
15:09 <dignifiedquire> It’s a mixture of misuse of bootstrap, and boostrap just being the wrong tool for the job in that case. Boostrap is a great thing and it makes a lot of things easier, but with modern browsers and flex layout you know have a basis where the browser actually brings most things you wanted from bootstrap in the first place. And for this electron app because we know we target only one single browser with excellent support for 
15:09 <dignifiedquire> these things you can write things just straight forward down in css. 
15:10 <•jbenet> customized styles to me, sound like moving away from this, and liable to (as the customized styles start piling up) a big mess where every person that comes in adds different rules in different ways making the whole thing much harder to maintain in general. (for this reason too, im pretty strict in go-ipfs when people start breaking common abstractions, and
15:10 <•jbenet> introducing one-offs that make things easy now, but harder to maintain in the long run).
15:11 ⇐ ygrek quit  •  edsu_ → edsu  
15:12 <•jbenet> hmmm. maybe i dont know what i'm talking about. you likely know this stuff better than me. i just worry about moving in a direction that's very hard to move out of, it's happened before in other parts of our codebases and it makes it very expensive in time to undo or to move through. (one example is the init/main part of go-ipfs. it's a quagmire)
15:12 <dignifiedquire> I just feel a whole ui framework like bootstrap is much harder to maintain in the long run and an unneeded abstraction. We could easily write a couple of guidelines and document the components we have to give people an easy way in to using it
15:13 <dignifiedquire> *them
15:14 ⇐ HostFat quit (~HostFat@2-228-102-98.ip191.fastwebnet.it) Read error: Connection reset by peer
15:14 <•jbenet> ok maybe propose it that way, and show how easy it would be for someone else to come in and use or add a new component to the mix. 
15:15 <dignifiedquire> Okay, I’ll do that, as soon as things are properly working
15:18 <dignifiedquire> jbenet: https://github.com/ipfs/electron-app/pull/43#issuecomment-145636652
15:18 <ion> Let's simply create an abstraction over bootstrap/flexbox! :-P
15:18 <•jbenet> ok sounds good -- the sooner the better as decreases time investment in that direction (i really hate having to be like "hmmm this doesn't work, we need to redo in some other way" it sucks for everyone)
15:19 ⇐ rendar quit (~I@host19-177-dynamic.42-79-r.retail.telecomitalia.it) Ping timeout: 256 seconds
15:19 <•jbenet> hahahaah thanks dignifiedquire
15:19 <•jbenet> :)
```

**Comment by Dignifiedquire** [comment]
@diasdavid for me the file drop is broken in the `settings` branch as well as here. It seems to me it doesn't even emit any `drop-files` event on the tray. Do you have any idea what's going wrong?

**Comment by jbenet** [comment]
@Dignifiedquire the jellyfish looks _amazing_ cc @whyrusleeping

**Comment by whyrusleeping** [comment]
dat jellyfish. dude. what. I'm in love.

**Comment by whyrusleeping** [comment]
@Dignifiedquire I added you to the file-browser repo i was working on (set to private at the moment because it doesnt work and is quite ugly): https://github.com/protocol/file-browser

I had drag-drop working nicely, you might take a look there and see how its different

**Comment by diasdavid** [comment]
> @diasdavid for me the file drop is broken in the settings branch as well as here. It seems to me it doesn't even emit any drop-files event on the tray. Do you have any idea what's going wrong?

My guess is that you are using a older version of electron, which didn't have the drag-drop support, I remember chasing that same bug when I first added it. Did you do npm install again when moving to the settings branch? For ref: http://electron.atom.io/docs/v0.33.0/api/tray/#event-39-drop-files-39-os-x

**Comment by Dignifiedquire** [comment]
@diasdavid yes I tried with 0.33.2 and 0.33.6

**Comment by diasdavid** [comment]
That is very weird, for me it continues to work:

image

Did it work for you before in your machine?

**Comment by Dignifiedquire** [comment]
@diasdavid no it never worked as far as I can remember :( What os version do you have? I upgraded to the captain (10.11) a couple of days ago maybe sth changed on there?

**Comment by diasdavid** [comment]
@dignifiedquire I'm running El Capitan too (since July). Will be with @jbenet today and will try out on his machine.

**Comment by Dignifiedquire** [comment]
@diasdavid okay thanks, atm the `settings` branch is completely broken for me, even on a fresh clone with fresh node_modules :( (on two different machines)

**Comment by diasdavid** [comment]
I had tried a fresh clone too. The other thing I can think of, is asking which Node.js version you are using, I'm on `v4.1.1`

**Comment by Dignifiedquire** [comment]
Reference issue to electron: https://github.com/atom/electron/issues/3008

**Comment by Dignifiedquire** [comment]
Finished file uploading with native notifications, and cross platform on the main window:

ipfs-file-drop

**Comment by Dignifiedquire** [comment]
I got the packaging to work, but running the packaged app fails because of https://github.com/jbenet/node-subcomandante/issues/1

**Comment by Dignifiedquire** [comment]
Merged into master

**Comment by diasdavid** [comment]
woot! :)

**Comment by jbenet** [comment]
Exciting! :+1:  i cant try it atm, but hope to later tonight or this weekend

**Comment by simonv3** [comment]
This looks awesome. :+1:

**Pull request closed without merging**